test/integration/DelayedRecovery.js

```diff
@@ -0,0 +1,388 @@
+    auxiliaryWeb3.eth.abi.encodeParameters(
+      ['bytes32', 'address'],
+      [RECOERY_MODULE_DOMAIN_SEPARATOR_TYPEHASH, recoveryModuleAddress]
+    )
+  );
+}
+
+function hashRecoveryModuleRecoveryStruct(structTypeHash, prevOwner, oldOwner, newOwner) {
+  return auxiliaryWeb3.utils.keccak256(
+    auxiliaryWeb3.eth.abi.encodeParameters(
+      ['bytes32', 'address', 'address', 'address'],
+      [structTypeHash, prevOwner, oldOwner, newOwner]
+    )
+  );
+}
+
+function hashRecoveryModuleRecovery(recoveryModuleAddress, structTypeHash, prevOwner, oldOwner, newOwner) {
+  const recoveryStructHash = hashRecoveryModuleRecoveryStruct(structTypeHash, prevOwner, oldOwner, newOwner);
+
+  const domainSeparatorHash = hashRecoveryModuleDomainSeparator(recoveryModuleAddress);
+
+  return auxiliaryWeb3.utils.soliditySha3(
+    { t: 'bytes1', v: '0x19' },
+    { t: 'bytes1', v: '0x01' },
+    { t: 'bytes32', v: domainSeparatorHash },
+    { t: 'bytes32', v: recoveryStructHash }
+  );
+}
+
+function signRecovery(recoveryModuleAddress, structTypeHash, prevOwner, oldOwner, newOwner, recoveryOwnerPrivateKey) {
+  const recoveryHash = hashRecoveryModuleRecovery(recoveryModuleAddress, structTypeHash, prevOwner, oldOwner, newOwner);
+
+  const signature = EthUtils.ecsign(EthUtils.toBuffer(recoveryHash), EthUtils.toBuffer(recoveryOwnerPrivateKey));
+
+  return {
+    recoveryHash,
+    signature
+  };
+}
+
+function signInitiateRecovery(recoveryModuleAddress, prevOwner, oldOwner, newOwner, recoveryOwnerPrivateKey) {
+  const INITIATE_RECOVERY_STRUCT_TYPEHASH = auxiliaryWeb3.utils.keccak256(
+    'InitiateRecoveryStruct(address prevOwner,address oldOwner,address newOwner)'
+  );
+
+  return signRecovery(
+    recoveryModuleAddress,
+    INITIATE_RECOVERY_STRUCT_TYPEHASH,
+    prevOwner,
+    oldOwner,
+    newOwner,
+    recoveryOwnerPrivateKey
+  );
+}
+
+function sleep(ms) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+async function waitBlockNumber(blockNumber) {
+  let currentBlockNumber = await auxiliaryWeb3.eth.getBlockNumber();
+
+  while (currentBlockNumber < blockNumber) {
+    // eslint-disable-next-line no-await-in-loop
+    await sleep(200);
+    currentBlockNumber = await auxiliaryWeb3.eth.getBlockNumber();
+  }
+}
+
+async function initiateAndExecuteRecovery(
+  userFactoryInstance,
+  owners,
+  threshold,
+  recoveryOwner,
+  recoveryControllerAddress,
+  recoveryBlockDelay,
+  prevOwnerAddress,
+  oldOwnerAddress,
+  newOwnerAddress
+) {
+  const { gnosisSafeProxyAddress } = await createUserWallet(
+    userFactoryInstance,
+    owners,
+    threshold,
+    recoveryOwner.address,
+    recoveryControllerAddress,
+    recoveryBlockDelay
+  );
+
+  const gnosisSafeProxy = instantiateGnosisSafeProxy(gnosisSafeProxyAddress);
+
+  const txOptions = {
+    from: deployerAddress,
+    gasPrice: ConfigReader.gasPrice,
+    gas: ConfigReader.gas
+  };
+
+  const modules = await gnosisSafeProxy.methods.getModules().call(txOptions);
```

Comment:
Can we know delayedRecoveryProxy address through event e.g. through UserWalletCreated ?
else we need to expose getModules() method in GnosisSafe interaction layer.

Reply:
1. for now ok, but we should avoid having a logic function like `hashRecoveryModuleRecovery` l144 in the test files;
2. if there is not already a good EIP712 js library out there, then I am in favor of splitting the EIP712 signer out of mosaic.js
3. working on the JLP examples we will gives us better clarity on the options for handling the signers outside of mosaic.js or openst.js

Reply:
1. there is value in exposing `getModules` for sure; let's open a ticket for that
2. the transaction receipt should include `EnabledModule(module)` as an event under the proxy contract address of GnosisSafe when setting up https://github.com/OpenSTFoundation/safe-contracts/blob/development/contracts/base/ModuleManager.sol#L43

Reply:
Ticket created: https://github.com/OpenSTFoundation/openst.js/issues/111